### PR TITLE
[Network] Add bitswap API

### PR DESCRIPTION
**Pull request by smnzhu**
Introduces new `BlockExchange` API, which is an abstraction around Bitswap.

This will be for state sync engine, among other things.

### TODO
- [x] Add tests

### CODE REVIEW

**Review by huitseeker — COMMENTED** [review]
Do we want to right away constrain the the payload on engine/execution/computation/computer/uploader's `BlockData` ( a struct that should probably live somewhere else in the hierarchy ), instead of reusing bitswap's "block"? In other terms, should we define an interface that hides the go-ipfs "Block" from a codebase that already has semantics for that word?

Another core TODO seems to be mapping those flow structures to CIDs.

**Comment by smnzhu** [comment]
> Do we want to right away constrain the the payload on engine/execution/computation/computer/uploader's `BlockData` ( a struct that should probably live somewhere else in the hierarchy ), instead of reusing bitswap's "block"? In other terms, should we define an interface that hides the go-ipfs "Block" from a codebase that already has semantics for that word?
> 
> Another core TODO seems to be mapping those flow structures to CIDs.

That depends on whether we foresee other usecases in the future for bitswap. I think there could be, and so the networking API should remain as general as possible. If you think the name Block is confusing, I can implement a wrapper interface around that instead and name it something else.

As long as there is the possibility that the state sync engine will not be the only usecase for bitswap, then I think all the application-specific logic of how to map BlockData to CID, how to sync an entire MerkleDAG and piece it back together into a BlockData, etc. should live in a separate module build on top of the network layer.

See for example https://github.com/onflow/flow-go/issues/1451 and https://github.com/onflow/flow-go/issues/1452

File: network/test/block_exchange_test.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+		ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(5*time.Second))
+
+		blocksReceived := make(map[cid.Cid]struct{})
+		done, err := bex.GetBlocks(blocksToGet...).ForEach(func(b blocks.Block) {
+			blocksReceived[b.Cid()] = struct{}{}
+		}).Send(ctx)
+		require.NoError(suite.T(), err)
+
+		<-done
```
Comment (vishalchangrani) [inline network/test/block_exchange_test.go:91]: nit: maybe use the unittest utility to time out if channel does not close within duration

File: network/test/block_exchange_test.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+		for j, blockCid := range suite.blockCids {
+			if j != i {
+				blocksToGet[blockCid] = struct{}{}
+			}
+		}
+
+		ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(5*time.Second))
+
+		var doneChans []<-chan struct{}
+		session := bex.GetSession(ctx)
```
Comment (vishalchangrani) [inline network/test/block_exchange_test.go:114]: what does a session signify here? Is this is a BlockSwap concept?

**Review by vishalchangrani — APPROVED** [review]
Left u a nit and a question but other than that lgtm.

File: network/test/block_exchange_test.go (hunk 2 above)
Comment (smnzhu) [reply network/test/block_exchange_test.go:114]: Yes, it is a bitswap concept: https://github.com/ipfs/go-bitswap#get-related-blocks-faster-with-sessions

**Comment by smnzhu** [comment]
bors merge

**Comment by smnzhu** [comment]
bors merge

**Comment by smnzhu** [comment]
bors merge

**Pull request merged**